src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala

```diff
@@ -446,6 +447,26 @@ trait V2Formats {
           )
         }
       }
+  }.map(addHealthCheckPortIndexIfNecessary)
+
+  /**
+    * Ensure backwards compatibility by adding portIndex to health checks when necessary.
+    *
+    * In the past, healthCheck.portIndex was required and had a default value 0. When we introduced healthCheck.port, we
+    * made it optional (also with ip-per-container in mind) and we have to re-add it in cases where it makes sense.
+    */
+  private[this] def addHealthCheckPortIndexIfNecessary(v2App: V2AppDefinition): V2AppDefinition = {
+    val hasPortMappings = v2App.container.exists(_.docker.exists(_.portMappings.exists(_.nonEmpty)))
+    val portIndexesMakeSense = v2App.ports.nonEmpty || hasPortMappings
+    if (portIndexesMakeSense) {
+      v2App.copy(healthChecks = v2App.healthChecks.map { healthCheck =>
+        if (healthCheck.port.isEmpty && healthCheck.portIndex.isEmpty && healthCheck.protocol != Protocol.COMMAND) {
```

Comment:
Can you extract those checks as flag: `val needPortIndex = x&y&z ` and then have only one if: `if (needPortIndex && portIndexesMakeSense) app.copy(...)`